.github/workflows/build-package.yml

```diff
@@ -38,18 +38,30 @@ jobs:
         password: ${{ secrets.GITHUB_TOKEN }}
 
     steps:
+    - name: Get correct ref
+      run: echo "GH_REF=${{ github.event_name == 'pull_request' }} && ${{ github.head_ref }} || ${{ github.ref_name }}" >> $GITHUB_ENV
+    
+    - name: Activate environment
+      run: spack env activate ${{ inputs.package-name }}
+    
     - name: Build package via spack
+      run: spack -d install --add --only package --fail-fast --no-checksum ${{ inputs.package-name }}@=$GH_REF%${{ inputs.compiler-name }}@${{ inputs.compiler-version }}
+    
+    - name: Generate lockfiles on failure
+      if: failure()
       run: |
-        spack -d install --only package --no-checksum ${{ inputs.package-name }}@=$GITHUB_REF_NAME%${{ inputs.compiler-name }}@${{ inputs.compiler-version }}
-      if: github.event_name != 'pull_request'
+        cd /opt/spack/var/spack/environments/${{ inputs.package-name }}
+        cp spack.lock ${{ inputs.package-name }}.old.spack.lock
+        spack -d concretize --force
+        cp spack.lock ${{ inputs.package-name }}.force.spack.lock
 
-    # Workaround for this issue: https://github.com/github/docs/issues/15319
-    # In the case of a pull request, we need to use GITHUB_HEAD_REF to get the
-    # correct branch name
-    - name: Build package via spack
-      run: |
-        spack -d install --only package --no-checksum ${{ inputs.package-name }}@=$GITHUB_HEAD_REF%${{ inputs.compiler-name }}@${{ inputs.compiler-version }}
-      if: github.event_name == 'pull_request'
+    - name: Upload lockfiles on failure
+      uses: actions/upload-artifact@v3
+      if: failure()
+      with: 
+        name: failed-lockfile-output
+        path: /opt/spack/var/spack/environments/${{ inputs.package-name }}/${{ inputs.package-name }}.*.spack.lock
+        
```

Comment:
We're uploading logs of failed builds for debugging and so they don't get deleted when workflow logs are cleaned?

Reply:
Not the logs, just the lockfiles. I think it'd be useful to see what the difference is between the lockfile that we have, and the lockfile that it regenerates to install it correctly, so we could update dependencies

Reply:
Agreed. I don't know if "logs" was a typo on my part or auto-correct, but I did understand they were lockfiles.